src/Interpreters/ExpressionActions.h

```diff
@@ -119,9 +118,29 @@ class ExpressionActions
     ExpressionActionsPtr clone() const;
 
 private:
+    struct Data
+    {
+        const Node * node = nullptr;
+        size_t num_created_children = 0;
+        std::vector<const Node *> parents;
+
+        ssize_t position = -1;
+        size_t num_created_parents = 0;
+        bool used_in_result = false;
+    };
+
```

Comment:
This struct is an implementation detail, and should not be shown in header file. (Sometimes, it is ok do declare it like `struct Data;` in .h and `struct ExpressionActions::Data { ... };` in .cpp).

What's better to do:
* Declare specific struct `ActionsDAGReverseInfo` or something like that (in .cpp, in anonymous namespace):
```cpp
struct ActionsDAGReverseInfo
{
    struct NodeInfo
    {
        std::vector<const Node *> parents;
        bool used_in_result = false;
    };

    using ReverseIndex = std::unordered_map<const Node *, size_t>>;
    ReverseIndex reverse_index;
};
```
* `std::pair<std::vector<Data>, std::unordered_map<const Node *, size_t>> getDataAndReverseIndex()` -> `static ActionsDAGReverseInfo getActionsDAGReverseInfo(const ActionsDAG & actions_dag);`; move it to .cpp.
* In `ExpressionActions::linearizeActions`, store `node, num_created_children, position, num_created_parents` in a separate `vector<Data>`